### ISSUE TITLE
Refactor of snmp_login to improve speed, detect "read-any" communities, improve "read-write" detection

### ISSUE THREAD

**Issue by hmoore-r7** [rapid7/metasploit-framework#5459]

**Comment by hmoore-r7** [comment]
This is a continuation of #4718. The existing code is slow and buggy.

**Comment by hmoore-r7** [comment]
Snagging this

**Comment by jsherwood0** [comment]
(Copied from reply in previous issue)

Thanks for the reply.  If I may ask, what was the problem with the module
before the refactor?  I found it to work well (other than a minor problem
with the progress indicators).  In fact, I often got my best results with
it.  Maybe I can retool the original to meet the needs that prompted the
refactor.

**Comment by hmoore-r7** [comment]
Quoting from the previous issue:
> The module before the refactor had many of the same problems as the current implementation (use of SNMP::Manager vs raw packets). The increased delay is the result of porting the module to the new LoginScanner model, which assumes a new connection for each login attempt. This doesn't work well for UDP services. It wasn't great before, it has been in this state for about six months, anyone with performance requirements around SNMP should look elsewhere until we can refactor this.

**Comment by jsherwood0** [comment]
OK.  I spent a bit of time looking at the current implementation, and have submitted a pull request with the fix (PR #5477).  

The problem seems to be that the new design (using `LoginScanner::SNMP` and `SNMP::Manager`) needed to take advantage of metasploit multithreading to be efficient, but it could not because it was configured to use `run_batch` approach to scanning.  (The `run_batch` setup was used in the original script to allow one batch to be kicked off for each community string, but the new design abandoned that since the `LoginScanner::SNMP` class takes a list of communities and iterates over that list for a given IP.)  By simply retooling the module to use `run_host` instead, the module can now take advantage of built-in multithreading and sees a huge performance increase.  Here are the timings that I gathered using the linux `time` command and a resource script to automate all setup and execution:

===Original Module===
real	10m49.583s
user	0m42.448s
sys	0m13.464s

===Refactored Module===
real	1174m9.721s
user	4m52.516s
sys	0m20.900s

===Fixed Module===
real	10m34.368s
user	4m21.352s
sys	0m19.472s


**Note:** All timings were done with THREADS=128 since most time is spent waiting for responses and not in CPU intensive activities.
